File: kernel/drivers/input/input.c
#include "../../../include/bazzulto/input.h"
#include "../../../include/bazzulto/waitqueue.h"
#include "../../../include/bazzulto/scheduler.h"
#include "../../../include/bazzulto/hal/hal_uart.h"

// ---------------------------------------------------------------------------
// Input ring buffer — single-producer (IRQ handlers) / single-consumer
// (input_getchar). All sources of stdin write here via input_emit_char().
// ---------------------------------------------------------------------------

#define INPUT_RING_SIZE 64

static char          input_ring[INPUT_RING_SIZE];
static volatile uint32_t input_ring_head;  // ISR writes here
static volatile uint32_t input_ring_tail;  // reader reads here
static wait_queue_t  input_wait_queue = WAIT_QUEUE_INIT;

void input_init(void)
{
    input_ring_head = 0;
    input_ring_tail = 0;
}

// Called from any IRQ handler that produces a character (keyboard, UART, etc.).
void input_emit_char(char character)
{
    // Ctrl+C (0x03): send SIGINT to the foreground process and echo "^C".
    // Do not place the character in the ring buffer; it is a control signal,
    // not data the reading process should see.
    if (character == 0x03) {
        hal_uart_putc('^');
        hal_uart_putc('C');
        hal_uart_putc('\r');
        hal_uart_putc('\n');
        scheduler_send_signal_to_foreground(2 /* SIGINT */);
        return;
    }

    uint32_t next_head = (input_ring_head + 1) % INPUT_RING_SIZE;
    if (next_head == input_ring_tail)
        return;  // ring full — drop character (preferred over blocking in IRQ context)

    input_ring[input_ring_head] = character;
    input_ring_head = next_head;

    process_wakeup(&input_wait_queue);
}

// Called by the VFS console read path. Blocks until a character is available.
// Returns the character (0–255) on success, or -1 if a signal is pending
// (so the caller can return from the syscall and deliver the signal).
int input_getchar(void)
{
    // Disable IRQs to prevent a race between checking the ring and going to
    // sleep — the same pattern used by uart_getc().
    __asm__ volatile("msr daifset, #2");

    while (input_ring_head == input_ring_tail) {
        // Check for pending signals before sleeping again.
        // Without this, a process blocked on console input would never see
        // a signal (e.g. SIGINT from Ctrl+C) because signals are only
        // delivered when the syscall handler returns to EL0.
        process_t *proc = scheduler_get_current();
        if (proc && proc->pending_signals) {
            __asm__ volatile("msr daifclr, #2");
            return -1;  // interrupted — let the syscall layer deliver the signal
        }
        process_sleep(&input_wait_queue);
        // process_sleep yields and returns when woken. Re-disable IRQs for
        // the next loop iteration.
        __asm__ volatile("msr daifset, #2");
    }

    char character = input_ring[input_ring_tail];
    input_ring_tail = (input_ring_tail + 1) % INPUT_RING_SIZE;

    __asm__ volatile("msr daifclr, #2");
    return (unsigned char)character;
}

File: kernel/src/drivers/input.rs
// drivers/input.rs — Input ring buffer abstraction.
//
// Single-producer (IRQ handlers) / single-consumer (sys_read on TTY) ring
// buffer. All character sources (keyboard, UART) write here via emit_char().
//
// Port of kernel/drivers/input/input.c.

use core::sync::atomic::{AtomicU32, Ordering};

const INPUT_RING_SIZE: usize = 64;

/// SIGINT signal number.
const SIGINT: u8 = 2;

// ---------------------------------------------------------------------------
// Ring buffer state
// ---------------------------------------------------------------------------

struct InputRing {
    buffer: [u8; INPUT_RING_SIZE],
    head: AtomicU32, // ISR writes here
    tail: AtomicU32, // reader reads here
}

impl InputRing {
    const fn new() -> Self {
        Self {
            buffer: [0u8; INPUT_RING_SIZE],
            head: AtomicU32::new(0),
            tail: AtomicU32::new(0),
        }
    }
}

struct SyncInputRing(core::cell::UnsafeCell<InputRing>);
unsafe impl Sync for SyncInputRing {}

static RING: SyncInputRing = SyncInputRing(core::cell::UnsafeCell::new(InputRing::new()));

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

pub fn init() {
    let ring = unsafe { &mut *RING.0.get() };
    ring.head.store(0, Ordering::Relaxed);
    ring.tail.store(0, Ordering::Relaxed);
}

/// Called from IRQ handlers. Enqueues a char; Ctrl+C → SIGINT to foreground.
pub fn emit_char(character: u8) {
    if character == 0x03 {
        // Ctrl+C: send SIGINT to foreground process, do not enqueue.
        unsafe {
            crate::drivers::uart::puts("^C\r\n");
            send_sigint_to_foreground();
        }
        return;
    }

    let ring = unsafe { &mut *RING.0.get() };
    let head = ring.head.load(Ordering::Relaxed);
    let next_head = (head + 1) % INPUT_RING_SIZE as u32;

    if next_head == ring.tail.load(Ordering::Acquire) {
        return; // ring full — drop character
    }

    ring.buffer[head as usize] = character;
    ring.head.store(next_head, Ordering::Release);
}

/// Called from sys_read on TTY. Blocks until a character is available.
/// Returns the character on success.
pub fn getchar() -> u8 {
    let ring = unsafe { &*RING.0.get() };

    // Disable IRQs to prevent race between checking ring and sleeping.
    unsafe { core::arch::asm!("msr daifset, #2", options(nostack, nomem)) };

    loop {
        let head = ring.head.load(Ordering::Acquire);
        let tail = ring.tail.load(Ordering::Relaxed);

        if head != tail {
            break;
        }

        // Yield to let other processes run while waiting.
        unsafe {
            core::arch::asm!("msr daifclr, #2", options(nostack, nomem));
            crate::scheduler::with_scheduler(|scheduler| {
                scheduler.schedule();
            });
            core::arch::asm!("msr daifset, #2", options(nostack, nomem));
        }
    }

    let tail = ring.tail.load(Ordering::Relaxed);
    let character = ring.buffer[tail as usize];
    ring.tail.store((tail + 1) % INPUT_RING_SIZE as u32, Ordering::Release);

    unsafe { core::arch::asm!("msr daifclr, #2", options(nostack, nomem)) };

    character
}

unsafe fn send_sigint_to_foreground() {
    unsafe { crate::scheduler::with_scheduler(|scheduler| {
        for slot_index in 0..crate::scheduler::PID_MAX {
            let pid = crate::process::Pid::new(slot_index as u16, 1);
            if let Some(process) = scheduler.process(pid) {
                if process.is_foreground {
                    let foreground_pid = process.pid;
                    scheduler.send_signal_to(foreground_pid, SIGINT);
                    return;
                }
            }
        }
    }); }
}
